### PR TITLE
Fix issue in block merging that causes fusing of separate tree instances. Fixes #28

### LINKED ISSUE #28: Block merging appears to erroneously fuse tree instances located far away from each other on sparse forest plots
Hi,

We were training SegmentAnyTree on a custom data set and obtained suspiciously low accuracy metrics for data that should be extremely easy to segment. Closer inspection revealed that the model had merged some trees that were located far away from each other (see picture below). The same problem occurs both with the provided pretrained weights and if the model is trained from scratch, so I don’t think the issue stems from the data itself. The integers in the picture are predicted instance ids and two trees with the same id are part of the same instance. Notice how multiple trees located far away from each other have been assigned to the same instance (3, 5, 7, 9).

Image

After some investigating, I narrowed the problem down to the block merging. As far as I understand, the following lines (`453–454` and `477–488` in `torch_points3d/metrics/panoptic_tracker_pointgroup_treeins.py`) cause an off by one error in some edge cases, which may result in merging instances unintentionally:
```
all_pre_ins[new_not_old_idx] = max_instance + 1
max_instance = max_instance + 1
```
Specifically, this seems to cause a problem in cases where no instances are merged when two blocks are merged (This is of course quite rare, unless the forest is very sparse. Even in our data set most of the data is significantly denser forest than the problem plot visualized above, so we didn’t even notice the issue at first). From what I gather, since `max_instance` is incremented before its current value has been assigned to any instance, the merging skips a label in this edge case. As a result, two instances may be assigned the same label, which results in the kind of instances shown in the attached image. The problem can be fixed e.g. by changing the lines to:
```
all_pre_ins[new_not_old_idx] = max_instance
max_instance = max_instance + 1
```
The segmentation results after implementing the fix are shown in the image below. Notice how the trees that were previously erroneously merged are now in their own segments.

Image

On our custom data set (quite similar to FOR-Instance in terms of point density), implementing these changes had the following effect on instance segmentation accuracy metrics (NOTE: model was trained from scratch on the custom data set):

Original metrics for test set:
* Precision: 67.81%
* Recall: 60.04%
* F1-score: 63.69% 

Metrics after implementing the change:
* Precision: 67.39%
* Recall: 60.30%
* F1-score: 63.63%

For the specific plot shown in the image, the metrics are as follows

Original metrics for the plot:
* Precision: 83.33%
* Recall: 62.50%
* F1-score: 71.43%

Metrics after implementing the change:
* Precision: 80.0%
* Recall: 100.0%
* F1-score: 88.90%

Across the entire data set, there’s a slight decrease in precision and slight increase in recall, while on the problem plot the issue appears to have been fixed. Based on these results, I would assume that the change does what was intended without any unintended side effects. Keep in mind that I’m not extremely familiar with the codebase since there is quite a bit of code, i.e. I may have misunderstood something. However, based on the accuracy metrics and qualitative results, I believe that this is indeed a bug, but please let me know if I’m incorrect in my assumption. Thanks!

### PR DESCRIPTION
**Pull request by ruoppa**
This pull request fixes a small bug in the block merging algorithm that caused tree instances located far away from each other to be fused in certain edge cases. This issue appears to only occur on relatively sparse forest plots. See #28 for a more detailed description of the problem.

### CODE REVIEW

**Comment by maciekwielgosz** [comment]
Thanks a lot for finding a bug and correction !
